### PR TITLE
feat(utilities): limb-level u256 serialization fast path + ethereum-executor intx->boost migration

### PR DESCRIPTION
**Pull request by kyonRay**
### Motivation

The repo's `toBigEndian`/`fromBigEndian` (bcos-utilities/DataConvertUtility.h) convert u256/u160 through a per-byte loop that runs one full 256-bit shift per byte — ~500ns per 32-byte conversion. A per-call-site census of the transaction path counts ~20 such conversions per transaction (ABI codec, receipt/ledger codecs, and the implicit `u256<->h256` conversions in FixedBytes.h:179/199), making this the single largest u256-related cost item. Hex and decimal paths are intentionally untouched: boost already wins those.

### Changes

**1. Limb-level fast path (`feat(utilities)`)**

For the fixed-width unsigned multiprecision types (u160/u256/u512 and every `FixedBytes<N>::ArithType`) over contiguous byte containers, the conversion now reads/writes whole 64-bit limbs with `boost::endian` byte-order handling instead of looping 256-bit shifts. Dispatch is a C++20 concept (`detail::LimbReadable` + `detail::ContiguousByteInput`); everything outside it — `bigint`, builtin integers, non-contiguous ranges, and 32-bit-limb builds (MSVC has no `__int128`) — keeps the original loop, preserved verbatim in the `else` branch.

Call sites are unchanged. Semantics are bit-identical to the old loops, including the container-decides-width silent truncation on both directions.

**2. intx hygiene (`chore(intx)`)**

- transaction-executor's `HostContext.h` included `<intx/intx.hpp>` without using any intx symbol; removed, and the module's now-orphaned `find_package(intx REQUIRED)` with it.
- `OpTestReceiptFactory.h` existed as two near-identical copies; kept the bcos-evm copy and pointed the two opstack-executor tests at it.

**3. ethereum-executor: intx -> bcos::u256 migration (`refactor(ethereum-executor)`)**

The tree's arithmetic and interfaces are boost-native now: EthAccount balances, the gas-price helpers, blob gas price and the fee accounting all use `bcos::u256`, with `bcos::u512` widening where the original used `intx::umul`/`uint512` (the 512-bit balance guard, blob fee, fake_exponential's overflow check). Narrowing/truncation semantics were probed identical between the two libraries before the swap (unchecked fixed cpp_int truncates low bits exactly like intx). The `toIntxU256`/`toBcosU256` bridge pair is deleted; `toEvmcBE`/`fromEvmcBE` on the new limb fast path replace the evmc-boundary conversions, so the former per-call byte-conversion tax on `ethMaxGasPrice` & co. is gone entirely.

Deliberately out of scope: `EVMPrecompiles.cpp` stays a verbatim evmone port (upstream diffability), `SECP256K1N_OVER_2` is converted once from evmone's intx curve constant, and opstack-executor — which keeps driving bcos-evm's intx types until its planned detach — now hosts its own local `u256 -> intx` bridge.

Verification for this commit: `test-ethereum-state-smoke` (keccak / CREATE / CREATE2 / EIP-4844 blob-price / EIP-7702 recovery golden vectors), `opstack-executor-tests`, `opstack-executor-receipt-tests`, `bcos-evm-eth-tests`, `bcos-evm-opstack-tests`, `test-transaction-scheduler` (end-to-end EthereumTransition) and `test-bcos-utilities` all green on a clean Debug build; `eest-runner` compiles (EEST fixtures are fetched in CI only).

### Verification

- **A/B oracle tests** (`DataConvertFastPathTest.cpp`): the pre-change loops are kept verbatim in the test as the oracle; equivalence is asserted for u256/u160/u512 x 16 container sizes (0, partial-limb edges, exact widths, over-wide 40/64/72/80) x bytes/std::string/std::array/bytesConstRef/C-array, with boundary values plus 200 varied-width randoms per type. `static_assert`s pin that the fast path engages exactly on 64-bit-limb platforms, so a silent fallback to the slow loop is a compile error, not a slowdown.
- **Test runs**: `test-bcos-utilities` and `test-bcos-codec` fully green on macOS/clang/boost-1.88; a standalone byte-equivalence probe on Ubuntu/gcc-14/**boost 1.83** reported 0 divergences; `opstack-executor-receipt-tests` and `bcos-evm-opstack-tests` green after the test-helper dedup.
- **Benchmarks** (256 fixed vectors, median of 5 reps, per-op):

| platform | toBigEndian | fromBigEndian |
|---|---|---|
| M1 Max, Apple clang, boost 1.88 | 500.7ns -> **1.03ns** | 511.9ns -> **0.84ns** |
| Xeon 8378C, gcc 14, boost 1.83 | 424.6ns -> **3.09ns** | 580.3ns -> **3.32ns** |

Worth noting: the dual-platform requirement caught a real bug during development — the trait's width parameter was first declared `unsigned`, mismatching `cpp_int_backend`'s `std::size_t`; Apple clang leniently matched the partial specialization anyway, while GCC (correctly) never selected it, silently keeping the slow path on x86. Fixed to `std::size_t` with `static_assert`s to prevent recurrence.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

**Review by morebtcg — COMMENTED** [review]
Reviewed the u256 fast-path + intx→boost migration. Overall the direction is good (bit-identical semantics, A/B oracle tests). A few concerns below — mainly around static-init ordering, one const-stripping cast, and a missing explicit include.

File: ethereum-executor/EVMSupport.h [hunk 1]
```diff
@@ -270,7 +269,11 @@ inline intx::uint256 compute_blob_gas_price(
 // ---- EIP-7702 authorization handling (ported bcos-evm Eip7702Recover.h) ----
 
 /// Secp256k1's N/2 is the upper bound of an EIP-2 canonical signature's s value.
-inline constexpr auto SECP256K1N_OVER_2 = evmmax::secp256k1::Curve::ORDER / 2;
+/// Converted once from evmone's intx constant — the curve API is the only intx
+/// boundary left in this header. Runtime-initialized: do not read it from
+/// another static initializer.
+inline const bcos::u256 SECP256K1N_OVER_2 =
```
Comment (morebtcg) [inline ethereum-executor/EVMSupport.h:275]: 把 `SECP256K1N_OVER_2` 从 `constexpr` 改成运行时初始化的 inline 变量存在静态初始化顺序（SIOF）风险：这个 header 会被多个 TU 包含，如果在某个 TU 的动态初始化阶段（其他 inline 变量 / 静态对象构造）调用了 `processAuthorizationList` 或直接读取该常量，会读到尚未初始化的值。注释里虽然警告了“不要从其他 static initializer 读取”，但库头文件很难保证所有调用点都在 `main()` 之后。

建议改成惰性初始化，例如函数局部 `static const bcos::u256`（C++11 起线程安全），或在首次使用时再转换，避免任何动态初始化期读取的可能。

File: bcos-utilities/bcos-utilities/DataConvertUtility.h [hunk 2]
```diff
@@ -297,10 +399,43 @@ inline void toBigEndian(T _val, Out& o_out)
...
-        ret = (T)((ret << 8) | (byte)(typename std::make_unsigned<decltype(i)>::type)i);
-    return ret;
+    if constexpr (detail::LimbReadable<T> && detail::ContiguousByteInput<_In>)
+    {
+        using Limb = boost::multiprecision::limb_type;
+        constexpr size_t valueBytes = detail::FixedWidthUnsigned<T>::bits / 8;
+        constexpr size_t limbCount = (detail::FixedWidthUnsigned<T>::bits + 63) / 64;
+        T ret = (T)0;
+        auto& backend = ret.backend();
+        backend.resize(limbCount, limbCount);
```
Comment (morebtcg) [inline bcos-utilities/bcos-utilities/DataConvertUtility.h:409]: `backend.resize(limbCount, limbCount)` 的第二个参数是 `cpp_int_backend::resize(new_size, min_size)` 的 `min_size`（分配下限提示），与第一个参数相同容易让人误读为“用 limbCount 填充”。建议只调用 `backend.resize(limbCount)`，或加一行注释说明第二个参数只是分配提示、并非初始化值。

File: bcos-utilities/bcos-utilities/DataConvertUtility.h [hunk 3]
```diff
@@ -283,10 +366,29 @@ inline void toBigEndian(T _val, Out& o_out)
...
+    if constexpr (detail::LimbReadable<T>)
     {
-        T v = _val & (T)0xff;
-        o_out[i - 1] = (typename Out::value_type)(uint8_t)v;
+        size_t outSize = o_out.size();
+        if constexpr (sizeof(typename Out::value_type) == 1 && requires { std::data(o_out); })
+        {
+            if (outSize == detail::FixedWidthUnsigned<T>::bits / 8)
+            {
+                detail::storeBigEndian(reinterpret_cast<uint8_t*>(std::data(o_out)), _val);
```
Comment (morebtcg) [inline bcos-utilities/bcos-utilities/DataConvertUtility.h:376]: `reinterpret_cast<uint8_t*>(std::data(o_out))` 会静默剥离 `std::data` 返回的 const。若调用方传入 `bytesConstRef` 之类的只读视图（其 `operator[]` 返回 const），原实现会编译失败，而这里会通过 cast 向只读对象写入——把编译期错误变成了静默 UB。建议给 `Out` 增加约束（例如要求 `std::is_const_v<typename Out::value_type>` 为 false），让 const 容器在编译期被拒绝，而不是靠调用方自觉。

File: ethereum-executor/tests/TestStorageBridge.h [hunk 4]
```diff
@@ -20,13 +20,28 @@
...
 #include <evmc/evmc.h>
 #include <optional>
 
 namespace bcos::test
 {
 
+// bcos::u256 <-> intx::uint256 for the bcos-evm oracle types this bridge drives (the
+// evmone::state interfaces stay intx). ethereum-executor's own arithmetic and interfaces no
+// longer use intx, so these test-local copies retire together with the bcos-evm dependency.
+inline intx::uint256 testToIntxU256(bcos::u256 const& val)
```
Comment (morebtcg) [inline ethereum-executor/tests/TestStorageBridge.h:33]: `testToIntxU256` / `testToBcosU256` 使用了 `intx::uint256`、`intx::be::unsafe::load`、`intx::be::store`，但本头文件没有显式 `#include <intx/intx.hpp>`，目前是依赖 bcos-evm 的 state 头间接引入（IWYU 问题）。建议显式包含，避免将来 bcos-evm 头文件调整后这里静默编译失败。

**Review by ywy2090 — APPROVED** [review]
@kyonRay · Round 1 · PR #5516 · Reviewed commit: `87d70688aae70b03765cbe9379496c76bd37a7c4` · Base: `6c39fa90b7fc330407d247c55898df5382952c72`.



**APPROVE** — first full 5+1 on the limb fast path and the intx→u256 executor migration. Open items are SUGGESTION/NIT only.

Public `toBigEndian`/`fromBigEndian` plus executor fee/balance conversion is a consensus-adjacent serialization surface, so this used five dimensions plus an isolated blind pass.

Independent lead check of the reported blob-fee wrap: BASE already did `uint256(total_blob_gas) * maxFee` then add to a 512-bit total, and `intx::umul` on the run path. This pin preserves that split (validate 256-wrap, run u512). Not treated as a new CRITICAL.

## Additional checks

- Fast path vs copied pre-change loops: empty / partial-limb / exact / over-wide on u160/u256/u512.
- `toEvmcBE`/`fromEvmcBE` are compile-time 32-byte only.
- `transaction-executor` `find_package(intx)` removal matches a tree with no leftover intx symbols.
- `nodeChainId` `fromBigEndian<uint64_t>` equals last-8-BE, same as `fromBigEndian<u256>`+`uint64` cast.

Optional follow-ups (not blocking): leftover PUBLIC intx in `EVMSupport.h`; four BE bridges; by-value `toBigEndian`; 32-bit oracle tautology; empty/over-wide unique asserts; shared chainId helper; `#include <utility>`; path:line comments.

## Verification boundaries

- Evidence is `git show` on `87d70688aae70b03765cbe9379496c76bd37a7c4` only (dirty worktree unused).
- Not compiled locally. Did not re-run `test-bcos-utilities` / EEST.
- Did not execute a 32-bit-limb build.

## Closing

No open CRITICAL/IMPORTANT. Merge is not blocked by this review.

File: ethereum-executor/EVMSupport.h (hunk 1 above)
Comment (morebtcg) [inline ethereum-executor/EVMSupport.h:275]: This loses the constexpr guarantee the intx constant had and leaves a static-init-order footgun that only the comment defends against. Consider `inline const bcos::u256& secp256k1nOver2() { static const auto v = fromEvmcBE(...); return v; }` — same one-time cost, but initialization-order-safe by construction instead of by convention.

File: bcos-utilities/test/unittests/libutilities/DataConvertFastPathTest.cpp [hunk 5]
```diff
@@ -0,0 +1,212 @@
...
+            for (auto& byteValue : input)
+            {
+                byteValue = (uint8_t)rng();
+            }
+            BOOST_REQUIRE(fromBigEndian<T>(input) == referenceFromBigEndian<T>(input));
+
+            std::string inputString(input.begin(), input.end());
+            BOOST_REQUIRE(fromBigEndian<T>(inputString) == referenceFromBigEndian<T>(inputString));
+
+            bytesConstRef inputRef(input.data(), input.size());
```
Comment (morebtcg) [inline bcos-utilities/test/unittests/libutilities/DataConvertFastPathTest.cpp:138]: Two small coverage gaps vs. the claimed matrix: (1) bytesConstRef is only exercised at offset 0 of a vector — a `bytesConstRef(input.data() + 1, size)` case would prove alignment-independence of the limb memcpy path; (2) the std::array exact-width lane in checkToBigEndianEquivalence is u256-only — worth adding std::array<uint8_t,20>/std::array<uint8_t,64> so u160's partial-top-limb store path is covered through the FixedBytes-shaped container too.

**Review by morebtcg — APPROVED** [review]
LGTM. Verified the limb fast path is byte-for-byte equivalent to the old shift loops in both directions (endianness, u160 partial top limb, over-wide/empty/narrow truncation), the concept dispatch + static_asserts make a silent fallback a compile error, all cross-width access goes through memcpy (no aliasing/alignment issues), and the intx→boost migration preserves semantics (u512 widening at the right sites, truncation parity, evmc BE boundary). Two non-blocking nits left inline.
